File: Pragmatic_Scala/src/main/java/Chaptor04_WorkingWithObjects/Singleton.scala
package Chaptor04_WorkingWithObjects

/**
 * @Description TODO
 * @Author sherlock
 * @Date
 */
object Singleton extends App {
	
	println(MarkerFactory getMarker "blue")
	println(MarkerFactory getMarker "blue")
	println(MarkerFactory getMarker "red")
	println(MarkerFactory getMarker "red")
	println(MarkerFactory getMarker "green")
	
}

class Marker(val color: String) {
	
	println(s"Creating $this")
	override def toString: String = s"marker color $color"
}

object MarkerFactory {
	
	private val markers = scala.collection.mutable.Map (
		"red" -> new Marker("red"),
		"blue" -> new Marker("blue"),
		"yellow" -> new Marker("yellow")
	)
	
	def getMarker(color: String): Marker = {
		markers.getOrElseUpdate(color, new Marker(color))
	}
}

File: Scala_Cookbook/src/main/java/cgs/java/SingleTon.java
package cgs.java;

/**
 * @Description TODO
 * @Author sherlock
 * @Date
 */
public class SingleTon {

    private SingleTon() {
        // private
    }

    private static class SingleTonHoler {
        private static final SingleTon INSTANCE = new SingleTon();
    }

    public static SingleTon getInstance() {
        return SingleTonHoler.INSTANCE;
    }
}
